### ISSUE TITLE
Error in `load_model()` if only one factor is active

### ISSUE THREAD

**Issue by const-ae** [bioFAM/MOFA2#36]
I get the following error message when I try to load a model:

```
Error in `colnames<-`(`*tmp*`, value = vals) : 
  attempt to set 'colnames' on an object with less than two dimensions
> traceback()
8: stop("attempt to set 'colnames' on an object with less than two dimensions")
7: `colnames<-`(`*tmp*`, value = vals)
6: `colnames<-`(`*tmp*`, value = vals)
5: .set_expectations_names(object, entity = "factors", value)
4: `factors_names<-`(`*tmp*`, value = paste0("Factor", as.character(seq_len(object@dimensions[["K"]]))))
3: `factors_names<-`(`*tmp*`, value = paste0("Factor", as.character(seq_len(object@dimensions[["K"]]))))
2: subset_factors(object, which(r2 > 1e-04))
1: load_model(outfile)
```

I am pretty sure this is because `subset_factors()` accidentally converts the expectation matrix to a vector due to a missing `drop=FALSE` [here](https://github.com/bioFAM/MOFA2/blob/f06ac7aff9fa7d3862436b639297c4a49fcd3467/MOFA2/R/subset.R#L181). This causes a subsequent error in `.set_expectations_names()` [here](https://github.com/bioFAM/MOFA2/blob/f06ac7aff9fa7d3862436b639297c4a49fcd3467/MOFA2/R/set_methods.R#L533). The error disappears if I set `remove_inactive_factors = FALSE`.

(I am sorry that I couldn't provide a small reproducible example, but `reprex` gives weird errors in combination with `reticulate`.)

Best Regards,
Constantin

**Comment by gtca** [comment]
Thanks, thanks for a detailed issue description!
Please let us know if 2de38d1 fixes the problem.

**Comment by const-ae** [comment]
Thank you for the quick reply. https://github.com/bioFAM/MOFA2/commit/2de38d1762b9381365d1b5f7b4a8a4667d17acc5 fixes the original error.

However, there is a new error now:

```
> model <- load_model(outfile, remove_inactive_factors = TRUE)
2 factors were found to explain no variance and they were removed for downstream analysis. You can disable this option by setting load_model(..., remove_inactive_factors = F)
Error in rowSums(sapply(object@cache[["variance_explained"]]$r2_per_factor,  : 
  'x' must be an array of at least two dimensions
> traceback()
3: stop("'x' must be an array of at least two dimensions")
2: rowSums(sapply(object@cache[["variance_explained"]]$r2_per_factor, 
       function(e) rowSums(e, na.rm = TRUE)))
1: load_model(outfile, remove_inactive_factors = TRUE)
```


This is the code that produces the problematic model:
```
library(MOFA2)
library(reticulate)
mofa <- import("mofapy2")
set.seed(1)
Y <- matrix(rpois(n = 400 * 25, lambda = 3), nrow = 400, ncol =25)
MOFAobject <- create_mofa(list("view1" = Y))
model_opts <- get_default_model_options(MOFAobject)
model_opts$num_factors <- 3
model_opts$likelihoods <- "poisson"
MOFAobject <- prepare_mofa(
  object = MOFAobject,
  model_options = model_opts
)
outfile <- "/tmp/test3.hdf5"
MOFAobject.trained <- run_mofa(MOFAobject, outfile)
model <- load_model(outfile, remove_inactive_factors = TRUE)
```

**Comment by gtca** [comment]
Thanks! It is all caused by R's automatic type conversion, and the code wasn't handling it well when only one factor was left.
I made a few fixes, and with `c1e4310` it seems to finally work. Does it work for you?

**Comment by const-ae** [comment]
I think you forgot a `browser()` statement [here](https://github.com/bioFAM/MOFA2/blob/c1e4310dbd19fa0df6d27d5979b68e30b1557691/MOFA2/R/load_model.R#L313).

But besides that, it works great now. Thank you :)

**Comment by gtca** [comment]
That one is also fixed now with 32fb9cdb.
Thanks!
